src/metabase/query_processor/middleware/process_userland_query.clj

```diff
@@ -172,13 +171,16 @@
   [qp :- ::qp.schema/qp]
   (mu/fn [query :- ::qp.schema/query
           rff   :- ::qp.schema/rff]
-    (if-not (get-in query [:middleware :userland-query?])
+    (if-not (userland-query? query)
       (qp query rff)
       (let [query          (assoc-in query [:info :query-hash] (qp.util/query-hash query))
-            execution-info (query-execution-info query)
-            field-usages   (query->field-usages query)]
+            execution-info (query-execution-info query)]
         (letfn [(rff* [metadata]
-                  (let [result (rff metadata)]
+                  (let [preprocessed-query (:preprocessed_query metadata)
+                        ;; we only need the preprocessed query to find field usages, so make sure we don't return it
+                        result             (rff (dissoc metadata :preprocessed_query))
+                        field-usages       (field-usage/pmbql->field-usages
```

Comment:
This seem to be happening synchronously during query execution, I wonder if there is any sense into passing it somewhere to execute async? Especially if we add parsing SQL in there.

Reply:
wdym by parsing SQL?

Reply:
Well right now it get field usages from mbql queries, right? But not from SQL?

Reply:
yes, only mbql